tractor/_broadcast.py

```diff
@@ -0,0 +1,308 @@
+                # or bail out on its own (thus un-subscribing)
+                state.subs[key] = state.maxlen - 1
+
+                # this task was overrun by the producer side
+                task: Task = current_task()
+                raise Lagged(f'Task {task.name} was overrun')
+
+            state.subs[key] -= 1
+            return value
+
+        # current task already has the latest value **and** is the
+        # first task to begin waiting for a new one
+        if state.recv_ready is None:
+
+            if self._closed:
+                raise trio.ClosedResourceError
+
+            event = trio.Event()
+            state.recv_ready = key, event
+
+            # if we're cancelled here it should be
+            # fine to bail without affecting any other consumers
+            # right?
+            try:
+                value = await self._recv()
+
+                # items with lower indices are "newer"
+                # NOTE: ``collections.deque`` implicitly takes care of
+                # trucating values outside our ``state.maxlen``. In the
+                # alt-backend-array-case we'll need to make sure this is
+                # implemented in similar ringer-buffer-ish style.
+                state.queue.appendleft(value)
+
+                # broadcast new value to all subscribers by increasing
+                # all sequence numbers that will point in the queue to
+                # their latest available value.
+
+                # don't decrement the sequence for this task since we
+                # already retreived the last value
+
+                # XXX: which of these impls is fastest?
+
+                # subs = state.subs.copy()
+                # subs.pop(key)
+
+                for sub_key in filter(
+                    # lambda k: k != key, state.subs,
+                    partial(ne, key), state.subs,
+                ):
+                    state.subs[sub_key] += 1
+
+                # NOTE: this should ONLY be set if the above task was *NOT*
+                # cancelled on the `._recv()` call otherwise sibling
+                # consumers will be awoken with a sequence of -1
+                event.set()
+
+                return value
+
+            finally:
+                # Reset receiver waiter task event for next blocking condition.
+                # this MUST be reset even if the above ``.recv()`` call
+                # was cancelled to avoid the next consumer from blocking on
+                # an event that won't be set!
+                state.recv_ready = None
+
+        # This task is all caught up and ready to receive the latest
+        # value, so queue sched it on the internal event.
+        else:
+            seq = state.subs[key]
+            assert seq == -1  # sanity
+            _, ev = state.recv_ready
+            await ev.wait()
+
+            seq = state.subs[key]
+            assert seq > -1, f'Invalid sequence {seq}!?'
```

Comment:
True since currently we'll never awake sibling consumers if the first-task-to-.receive() is cancelled.